core/src/main/scala/org/apache/spark/RecoverCacheShutdown.scala

```diff
@@ -0,0 +1,234 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark
+
+import java.util.concurrent.{ScheduledFuture, TimeUnit}
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Failure
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.rpc.RpcEndpointRef
+import org.apache.spark.storage.{BlockId, BlockManagerId, RDDBlockId}
+import org.apache.spark.storage.BlockManagerMessages.{GetCachedBlocks, GetMemoryStatus, GetSizeOfBlocks, ReplicateOneBlock}
+import org.apache.spark.util.ThreadUtils
+
+/**
+ * Responsible for asynchronously replicating all of an executors cached blocks, and then shutting
+ * it down.
+ */
+final private class RecoverCacheShutdown(
+    state: RecoverCacheShutdownState,
+    conf: SparkConf)
+  extends Logging {
+
+  private val threadPool = ThreadUtils.newDaemonCachedThreadPool("recover-cache-shutdown-pool")
+  private implicit val asyncExecutionContext = ExecutionContext.fromExecutorService(threadPool)
+
+  /**
+   * Start the recover cache shutdown process for these executors
+   *
+   * @param execIds the executors to start shutting down
+   */
+  def startExecutorKill(execIds: Seq[String]): Unit = {
+    logDebug(s"Recover cached data before shutting down executors ${execIds.mkString(", ")}.")
+    checkForReplicableBlocks(execIds)
+  }
+
+  /**
+   * Stops all thread pools
+   *
+   * @return
+   */
+  def stop(): java.util.List[Runnable] = {
+    threadPool.shutdownNow()
+    state.stop()
+  }
+
+  /**
+   * Get list of cached blocks from BlockManagerMaster. If there are cached blocks, replicate them,
+   * otherwise kill the executors
+   *
+   * @param execIds the executors to check
+   */
+  private def checkForReplicableBlocks(execIds: Seq[String]) = state.getBlocks(execIds).foreach {
+    case (executorId, NoMoreBlocks) => state.killExecutor(executorId)
+    case (executorId, NotEnoughMemory) => state.killExecutor(executorId)
+    case (executorId, HasCachedBlocks) => replicateBlocks(executorId)
+  }
+
+  /**
+   * Replicate one cached block on an executor. If there are more, repeat. If there are none, check
+   * with the block manager master again. If there is an error, go ahead and kill executor.
+   *
+   * @param execId the executor to save a block one
+   */
+  private def replicateBlocks(execId: String): Unit = {
+    val (response, blockId) = state.replicateFirstBlock(execId)
+    response.onComplete {
+      case scala.util.Success(true) =>
+        logTrace(s"Finished replicating block ${blockId.getOrElse("unknown")} on exec $execId.")
+        replicateBlocks(execId)
+      case scala.util.Success(false) => checkForReplicableBlocks(Seq(execId))
```

Comment:
Could you import the types instead? Also the code after the `=>` would look better in the next line, given that's how the other cases do it.

Reply:
There was an issue with scala.util.Success getting shadowed by the success object in TaskEndReason.scala, but that's fixed.